File: fortran/src/cpf.f90
module cpf_module
!----------------------------------------------------------------------!
! This module provides 2 functions that compute complex probability
! function (CPF):
! 1. cpf_accurate - computes the CPF using rational series with 42 terms
! 2. cpf_fast - computes the CPF using Humlicek's algorithm and rational
!    series with 24 terms
!----------------------------------------------------------------------!
   use, intrinsic :: iso_fortran_env, only: int32, dp => real64
   implicit none
!----------------------------------------------------------------------!
   real(dp), parameter :: inverse_sqrt_pi = 0.5641895835477563_dp
!----------------------------------------------------------------------!
   contains
!----------------------------------------------------------------------!
   function cpf_accurate(x, y) result(cpf_z)
      !----------------------------------------------------------------!
      ! Computes the complex probability function using a rational
      ! series with 42 terms. It is assumed that Im(z) > 0 or Im(z) = 0.
      ! A series was simplified to 37 terms introducing less than
      ! 10^(-17) deviations on mHT profile.
      !
      ! Input/Output Parameters of Routine
      ! ---------------------------------------------------------------!
      ! x : Real part of input complex parameter
      ! y : Imaginary part of input complex parameter
      !
      ! The function provides one output:
      ! ---------------------------------------------------------------!
      ! (1): Complex probability function
      !----------------------------------------------------------------!
      real(dp), intent(in) :: x, y
      complex(dp) :: cpf_z
      !----------------------------------------------------------------!
      integer(int32), parameter :: number_of_fft_terms = 37
      real(dp), parameter :: weidemann_constant_42 = 5.449631621480024_dp
      real(dp), dimension(number_of_fft_terms) :: fft_constant_terms=(/&
         -3.129493160727961e-14_dp, -1.188364999909099e-14_dp,         &
          1.951777029849348e-13_dp,  1.790586243645278e-13_dp,         &
         -1.184560208678836e-12_dp, -2.069163661083667e-12_dp,         &
          6.430136110306704e-12_dp,  2.063579921011804e-11_dp,         &
         -2.392389527320517e-11_dp, -1.799169607159564e-10_dp,         &
         -6.353807951660892e-11_dp,  1.282896083944607e-9_dp,          &
          2.636162411919059e-09_dp, -5.468780625369738e-9_dp,          &
         -3.294773119114329e-8_dp,  -2.752070035718561e-8_dp,          &
          2.206733163926054e-7_dp,   8.511689670641750e-7_dp,          &
          4.936972061734341e-7_dp,  -6.617492208403963e-6_dp,          &
         -2.914574364851397e-5_dp,  -4.816473680511106e-5_dp,          &
          1.044072210002090e-4_dp,   1.070131083157417e-3_dp,          &
          4.631075611097791e-3_dp,   1.480296368764821e-2_dp,          &
          3.922970169744468e-2_dp,   9.038744880336540e-2_dp,          &
          1.857036333535562e-1_dp,   3.455278077566057e-1_dp,          &
          5.882708203344523e-1_dp,   9.230959991941070e-1_dp,          &
          1.342044484596932_dp,      1.814714451499866_dp,             &
          2.288734169675538_dp,      2.697763665856064_dp,             &
          2.975931371735470_dp/)
      !----------------------------------------------------------------!
      complex(dp) :: z, z_ratio
      integer(int32) :: i
      !----------------------------------------------------------------!
      z       = cmplx(-y, x, kind=dp)
      z_ratio = (weidemann_constant_42 + z) / (weidemann_constant_42 - z)
      cpf_z   = (0.0_dp, 0.0_dp)
      do i = 1, number_of_fft_terms
         cpf_z = cpf_z + fft_constant_terms(i)                         &
            * z_ratio**(real(number_of_fft_terms - i, kind = dp))
      end do
      cpf_z = 2.0_dp * cpf_z / (weidemann_constant_42 - z)**2.0_dp     &
         + inverse_sqrt_pi / (weidemann_constant_42 - z)
      !----------------------------------------------------------------!
   end function cpf_accurate
!----------------------------------------------------------------------!
   function cpf_fast(x, y) result(cpf_z)
      !----------------------------------------------------------------!
      ! Computes the complex probability function using Humlicek's 
      ! algorithm in its first subregion (Source: 10.1016/0022-4073(82)90078-4) 
      ! and using a rational series with 24 terms in other subregions.
      !
      ! Input/Output Parameters of Routine
      ! ---------------------------------------------------------------!
      ! x : Real part of input complex parameter
      ! y : Imaginary part of input complex parameter
      !
      ! The function has one output:
      ! ---------------------------------------------------------------!
      ! (1): Complex probability function
      ! ---------------------------------------------------------------!
      real(dp), intent(in) :: x, y
      complex(dp) :: cpf_z
      ! ---------------------------------------------------------------!
      integer(int32), parameter :: number_of_fft_terms = 24
      real(dp), parameter :: hum1_threshold = 15.0_dp
      real(dp), parameter :: weidemann_constant_24 = 4.119534287814236_dp
      real(dp), dimension(number_of_fft_terms) :: fft_constant_terms=(/&
         -1.513747622620502e-10_dp, 4.904820407381768e-9_dp,           &
          1.331045329581992e-9_dp, -3.008282344381996e-8_dp,           &
         -1.912225887484805e-8_dp,  1.873834346505099e-7_dp,           &
          2.568264135399530e-7_dp, -1.085647579417637e-6_dp,           &
         -3.038893184366094e-6_dp,  4.139461724429617e-6_dp,           &
          3.047106608295325e-5_dp,  2.433141546207148e-5_dp,           &
         -2.074843151143828e-4_dp, -7.816642995626165e-4_dp,           &
         -4.936426901286291e-4_dp,  6.215006362949147e-3_dp,           &
          3.372336685531603e-2_dp,  1.083872348456673e-1_dp,           &
          2.654963959880772e-1_dp,  5.361139535729116e-1_dp,           &
          9.257087138588670e-1_dp,  1.394819673379119_dp,              &
          1.856286499205540_dp,     2.197858936531542_dp/)
      !----------------------------------------------------------------!
      complex(dp) :: t, z, z_ratio
      integer(int32) :: i
      !----------------------------------------------------------------!
      if (abs(x) + y > hum1_threshold) then
         !-------------------------------------------------------------!
         t = cmplx(y, -x, kind=dp)
         cpf_z = inverse_sqrt_pi * t / (0.5_dp + t**2.0_dp)
         !-------------------------------------------------------------!
      else
         !-------------------------------------------------------------!
         z       = cmplx(-y, x, kind=dp)
         z_ratio = (weidemann_constant_24 + z) / (weidemann_constant_24 - z)
         cpf_z   = (0.0_dp, 0.0_dp)
         do i = 1, number_of_fft_terms
            cpf_z = cpf_z + fft_constant_terms(i)                      &
               * Z**(real(number_of_fft_terms - i, kind = dp))
         end do
         cpf_z = 2.0_dp * cpf_z / (weidemann_constant_24 - z)**2.0_dp  &
            + inverse_sqrt_pi / (weidemann_constant_24 - z)
         !-------------------------------------------------------------!
      endif
      !----------------------------------------------------------------!
  end function cpf_fast
!----------------------------------------------------------------------!
end module cpf_module

File: python/mHT/CPF.py
inverse_sqrt_pi = 0.5641895835477563

def cpf_accurate(x,y):
    """    
    # ----------------------------------------
    #      "CPF_ACCURATE": Accurate CPF algorithm
    #      Computes the complex probability function using a rational series 
    #      with 42 terms. It is assumed that Im(z) > 0 or Im(z) = 0. (reference: jstor.org/stable/2158232)
    #      A series was simplified to 37 terms introducing less than 10^(-17)
    #      deviations on mHT profile.
    #
    #      Standard Input Parameters:
    #      -------------------- 
    #      x  : Real part of input complex parameter
    #      y  : Imaginary part of input complex parameter
    #
    #      The function has one output:
    #      --------------------
    #      (1): Complex probability function
    # ----------------------------------------
    """
    z = -y + x*1j
    L = 5.449631621480024 # The pre-calculated Weideman constant for N = 42
    Z = (L+z)/(L-z)
    a = [ -3.129493160727961E-14, -1.188364999909099E-14,  1.951777029849348E-13,  1.790586243645278E-13, 
          -1.184560208678836E-12, -2.069163661083667E-12,  6.430136110306704E-12,  2.063579921011804E-11, 
          -2.392389527320517E-11, -1.799169607159564E-10, -6.353807951660892E-11,  1.282896083944607E-09,  
           2.636162411919059E-09, -5.468780625369738E-09, -3.294773119114329E-08, -2.752070035718561E-08,  
           2.206733163926054E-07,  8.511689670641750E-07,  4.936972061734341E-07, -6.617492208403963E-06, 
          -2.914574364851397E-05, -4.816473680511106E-05,  1.044072210002090E-04,  1.070131083157417E-03, 
           4.631075611097791E-03,  1.480296368764821E-02,  3.922970169744468E-02,  9.038744880336540E-02, 
           1.857036333535562E-01,  3.455278077566057E-01,  5.882708203344523E-01,  9.230959991941070E-01, 
           1.342044484596932E+00,  1.814714451499866E+00,  2.288734169675538E+00,  2.697763665856064E+00, 
           2.975931371735470E+00] # The pre-calculated table of FFT constant terms (truncated from the begining)
    p= a[36] + (a[35] + (a[34] + (a[33] + (a[32] + (a[31] + (a[30] + (a[29] + (a[28] + (a[27] + (a[26] + (a[25] + (a[24] + (a[23] + (a[22] + (a[21] + (a[20] + (a[19] + (a[18] + (a[17] + (a[16] + (a[15] + (a[14] + (a[13] +(a[12] +  (a[11] + (a[10] + (a[9] + (a[8] + (a[7] + (a[6] + (a[5] + (a[4] + (a[3] + (a[2] + (a[1] + a[0] * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z)* Z
    return (2*p/(L-z) + inverse_sqrt_pi)/(L-z)

def cpf_fast(x,y):
    """    
    # ----------------------------------------
    #      "CPF_FAST": Fast CPF algorithm
    #      Computes the complex probability function using Humlicek's 
    #      algorithm in its first subregion (Source: 10.1016/0022-4073(82)90078-4) 
    #      and using a rational series with 24 terms in other subregions. (reference: jstor.org/stable/2158232)
    #
    #      Standard Input Parameters:
    #      -------------------- 
    #      x  : Real part of input complex parameter
    #      y  : Imaginary part of input complex parameter
    #
    #      The function has one output:
    #      --------------------
    #      (1): Complex probability function
    # ----------------------------------------
    """
    hum1_threshold = 15.0 # the border of the first Humlicek's region
    if abs(x)+y>hum1_threshold:
        t = y-x*1j
        return inverse_sqrt_pi*t/(0.5+pow(t,2))
    else: 
        z = -y + x*1j
        L = 4.119534287814236 # the pre-calculated Weideman constant for N = 24
        Z = (L+z)/(L-z)
        a = [ -1.513747622620502E-10,  4.904820407381768E-09,  1.331045329581992E-09, -3.008282344381996E-08,
              -1.912225887484805E-08,  1.873834346505099E-07,  2.568264135399530E-07, -1.085647579417637E-06,
              -3.038893184366094E-06,  4.139461724429617E-06,  3.047106608295325E-05,  2.433141546207148E-05,
              -2.074843151143828E-04, -7.816642995626165E-04, -4.936426901286291E-04,  6.215006362949147E-03,
               3.372336685531603E-02,  1.083872348456673E-01,  2.654963959880772E-01,  5.361139535729116E-01,
               9.257087138588670E-01,  1.394819673379119E+00,  1.856286499205540E+00,  2.197858936531542E+00] 
            # the pre-calculated table of FFT constant terms
        p = a[23] + (a[22] + (a[21] + (a[20] + (a[19] + (a[18] + (a[17] + (a[16] + (a[15] + (a[14] + (a[13] +(a[12] +  (a[11] + (a[10] + (a[9] + (a[8] + (a[7] + (a[6] + (a[5] + (a[4] + (a[3] + (a[2] + (a[1] + a[0] * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z) * Z
        return (2*p/(L-z) + inverse_sqrt_pi)/(L-z)
